unstructured/ingest/v2/processes/connectors/salesforce.py

```diff
@@ -0,0 +1,293 @@
+
+        client = self.connection_config.get_client()
+
+        files_list = []
+        for record_type in self.index_config.categories:
+            if record_type not in ACCEPTED_CATEGORIES:
+                raise ValueError(f"{record_type} not currently an accepted Salesforce category")
+
+            try:
+                # Get ids from Salesforce
+                records = client.query_all_iter(
+                    f"select Id, SystemModstamp, CreatedDate, LastModifiedDate from {record_type}",
+                )
+                for record in records:
+                    files_list.append(
+                        FileData(
+                            connector_type=CONNECTOR_TYPE,
+                            identifier=record["Id"],
+                            source_identifiers=SourceIdentifiers(
+                                filename=record["Id"],
+                                fullpath=f"{record['attributes']['type']}/{record['Id']}",
+                            ),
+                            metadata=DataSourceMetadata(
+                                url=record["attributes"]["url"],
+                                version=record["SystemModstamp"],
+                                date_created=record["CreatedDate"],
+                                date_modified=record["LastModifiedDate"],
+                                record_locator={"id": record["Id"]},
+                            ),
+                            additional_metadata={"record_type": record["attributes"]["type"]},
+                        )
+                    )
+            except SalesforceMalformedRequest as e:
+                raise SalesforceMalformedRequest(f"Problem with Salesforce query: {e}")
+
+        return files_list
+
+    def run(self, **kwargs: Any) -> Generator[FileData, None, None]:
+        for f in self.list_files():
+            yield f
+
+
+@dataclass
+class SalesforceDownloaderConfig(DownloaderConfig):
+    pass
+
+
+@dataclass
+class SalesforceDownloader(Downloader):
+    connection_config: SalesforceConnectionConfig
+    download_config: SalesforceDownloaderConfig = field(
+        default_factory=lambda: SalesforceDownloaderConfig()
+    )
+    connector_type: str = CONNECTOR_TYPE
+
+    def get_file_extension(self, record_type) -> str:
+        if record_type == "EmailMessage":
+            extension = ".eml"
+        elif record_type in ["Account", "Lead", "Case", "Campaign"]:
+            extension = ".xml"
+        else:
+            raise MissingCategoryError(
+                f"There are no categories with the name: {record_type}",
+            )
+        return extension
+
+    def get_download_path(self, file_data: FileData) -> Path:
+        record_file = file_data.identifier + self.get_file_extension(
+            file_data.additional_metadata["record_type"]
+        )
+        return Path(self.download_dir) / file_data.additional_metadata["record_type"] / record_file
```

Comment:
This logic should be part of the code creating the source identifiers above (setting filename), and then that should be used when creating the download path. Otherwise the `source_identifiers.filename` isn't accurate since it doesn't contain the extension.

Reply:
@rbiseck3 I thought about this. there are only 2 potential extensions. .eml and .xml and they are only used to find a nice way to process the files we get from Salesforce. So it's actually not accurate to claim that a file is .xml . up for debate though.